### PR TITLE
[BEAM-1530] Support record-determined output tables

### PR DESCRIPTION
**Pull request by reuvenlax**
Changes the existing tableRefFunction: BoundedWindow -> String, to a policy that takes a Context object. This allows us to add further items to the policy without breaking compatibility. This PR puts the window and the element in the Context.

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @jkff 

Eugene, do you perhaps have a few cycles to review this one?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 1]
```diff
@@ -158,7 +157,6 @@ private static void populateCommonDisplayData(DisplayData.Builder builder,
    * a map of attached attributes.
    */
   public static class PubsubMessage {
-
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: These changes seem unrelated to the current PR.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 2]
```diff
@@ -209,21 +209,21 @@
  * <h3>Sharding BigQuery output tables</h3>
  *
  * <p>A common use case is to dynamically generate BigQuery table names based on
- * the current window. To support this,
- * {@link BigQueryIO.Write#to(SerializableFunction)}
+ * the current window or element. To support this,
+ * {@link BigQueryIO.Write#to(BigQueryIO.TableWritePolicy)}
  * accepts a function mapping the current window to a tablespec. For example,
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: current window or element?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 3]
```diff
@@ -209,21 +209,21 @@
...
  * here's code that outputs daily tables to BigQuery:
  * <pre>{@code
  * PCollection<TableRow> quotes = ...
  * quotes.apply(Window.<TableRow>into(CalendarWindows.days(1)))
  *       .apply(BigQueryIO.Write
  *         .withSchema(schema)
- *         .to(new SerializableFunction<BoundedWindow, String>() {
- *           public String apply(BoundedWindow window) {
+ *         .to(new TableWritePolicy<String>() {
+ *           public String apply(Context window) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: It's not a window. Please audit the rest of the file for other inconsistencies - I found too many to put individual comments on each.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 4]
```diff
@@ -434,6 +434,33 @@ public TableReference apply(String jobUuid) {
   }
 
   /**
+   * A policy for determining output table names. Currently supports examining the elemented or
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: the element

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 5]
```diff
@@ -1588,20 +1615,19 @@ public static Bound to(TableReference table) {
      * <p>{@code tableSpecFunction} should be deterministic. When given the same window, it should
      * always return the same table specification.
      */
-    public static Bound to(SerializableFunction<BoundedWindow, String> tableSpecFunction) {
-      return new Bound().to(tableSpecFunction);
+    public static Bound to(TableWritePolicy<String> tableRefPolicy) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: tableWritePolicy perhaps?

In general I'd suggest to rename the concept to something else - "table write policy" is overly generic. Maybe something with the word "routing" or "destination"? e.g. call it DestinationTableFn?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 6]
```diff
@@ -434,6 +434,33 @@ public TableReference apply(String jobUuid) {
   }
 
   /**
+   * A policy for determining output table names. Currently supports examining the elemented or
+   * the window.
+   */
+   public abstract static class TableWritePolicy<T> implements SerializableFunction<
+      TableWritePolicy.Context, T> {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Rename T to TableT or something like that? For a while I thought it refers to the type of the element.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 7]
```diff
@@ -1172,11 +1173,11 @@ public String apply(TableRow value) {
         }
     );
 
-    SerializableFunction<BoundedWindow, String> tableFunction =
-        new SerializableFunction<BoundedWindow, String>() {
+    TableWritePolicy<String> tableFunction = new TableWritePolicy<String>() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Suppose there should also be a RunnableOnService test that actually uses the element.

**Comment by reuvenlax** [comment]
BTW this change was done as a minimum change to get the functionality
needed (minor extension of existing functionality).

IMO a better API would change BQIO entirely as follows:

BigQueryIO.Write takes input of type Foo, and a SerializableFunction that
maps Foo -> TableRow. This would allow the table mapping to operate on Foo
instead of TableRow. It would also better open the door for convenience
mappers that could automatically generate a correct schema from Avro types
and convert the Avro type to a TableRow.

I refrained from implementing this earlier as it's a major
backwards-incompatible change, and would probably engender a degree of
bikeshedding. However it might be a better approach for BigQueryIO.Write in
general, so worth bringing up.

**Comment by jkff** [comment]
I guess in general this falls under https://issues.apache.org/jira/browse/BEAM-92 which @vikkyrk worked on for a while but it was deprioritized.

But while we don't have that API, let's discuss the other options. I think the issue with current approach in the PR - that you can't write same-looking TableRows to different tables - is serious enough that we should not submit it like this.

The other approach is what you suggest - let users map a Foo to TableRow and table name / schema. The addition of schema makes it less tempting to do the table name & schema computation for every Foo. Maybe something like this? (which is similar to some of the approaches we were considering for BEAM-92)

```
foos.apply(BigQueryIO.Write.to(BigQueryIO.Write.TableDispatcher<Foo, ?>))
```

where:

```
interface TableDispatcher<InputT, DestinationT> {
  DestinationT getDestination(InputT element);
  String getTableName(DestinationT destination);
  Schema getTableSchema(DestinationT destination);
  ... maybe other stuff like "create/write disposition" even? ...
  TableRow toTableRow(InputT element);
}
```

**Comment by reuvenlax** [comment]
BTW, are you trying to avoid having multiple builder methods to specify
each thing? The downside of this approach is that the user must always
specify all methods. In the (common) case where all tables have the same
schema, this is harder to use.

Your approach allows for per-table schemas as well, which goes a bit
further than what my PR did. If we do go with this, I believe
getTableSchema needs a way of accessing a list of PCollectionView objects
as well; we've had several user requests for this in batch - a previous
stage calculates the schema based on the entire input PCollection.

**Comment by jkff** [comment]
Interesting. I think in general, and even closer to BEAM-92, it could look roughly like this then:

```
foos.apply(BigQueryIO.Write.to(
    (foo, window) -> destination,
    (destination) -> new BigQueryIO.Write.Destination(...table name, schema, other options...),
    (foo) -> new TableRow(...))
```

Encapsulating properties of a destination table in a POJO is better than representing them as virtual functions.

This requires that all of these functions, here represented as lambdas, should statically declare and be able to access side inputs. This could be done by shoehorning it into this BigQueryIO.Write.to() API, or could be done by adding an abstraction like SerializableFunctionWithSideInputs or alike.

**Comment by reuvenlax** [comment]
Sure. Also if destination is a POJO (v.s. a string), we can then add a
SerializableFunctionWithSideInputs version later without breaking backwards
compatibility. This allows us to postpone that functionality for now.

**Comment by jkff** [comment]
OK, so are you planning to go ahead with this suggestion or something similar in the current PR?

**Comment by reuvenlax** [comment]
Will send an RFC pr later this morning.
